### PR TITLE
Continue indexing on deterministic error

### PR DESCRIPTION
**Pull request by leoyvens**
This implements the non-fatal errors proposal, by which a subgraph which has synced will continue indexing when encountering a deterministic error, skipping the handler and discarding any changes done while processing it. The errors will be registered when transacting operations for a block, and will be reverted if necessary. To enable non-fatal errors the manifest must include the section:
```
features:
  - nonFatalErrors
```

By default data is hidden from queries in the presence of indexing errors (except for the `_meta` field). To receive data in a query the `subgraphError: allow` argument must be passed in. See the `non_fatal_errors` test for examples of what the responses look like. The `_meta` field gains an `hasIndexingErorrs: Boolean!` field as a way of checking for errors. Additionally if indexing errors are present a graphql error with the message `"indexing_error"` will always be added.

This is ready for review. The `nonFatalErrors` field of the indexing status API will be implemented in a follow up. I'd also like to add an integration test before merging this.

### CODE REVIEW

File: store/postgres/src/metadata.rs
Comment (lutter) [inline store/postgres/src/metadata.rs]: this should use `block_range::BLOCK_RANGE_COLUMN` as the name of the block range column

File: graph/src/components/store.rs
Comment (lutter) [inline graph/src/components/store.rs:1285]: Similarly, let's please hold off on asyncifying store operations.

File: graph/src/data/query/error.rs
Comment (lutter) [inline graph/src/data/query/error.rs:255]: It seems weird that any `anyhow::Error` gets translated into a `QueryExecutionError::StoreError`

File: graphql/src/query/ext.rs
Comment (lutter) [inline graphql/src/query/ext.rs:85]: Nice! No idea what drugs I was on when I initially wrote that code.

File: graphql/src/schema/api.rs
Comment (lutter) [inline graphql/src/schema/api.rs:219]: Not sure if you noticed, but I added a `meta.graphql` file a while ago. Would be cleaner to add this type definition to that file, too, since a GraphQL schema definition is much easier to read than Rust code that builds the AST.

File: graphql/tests/query.rs
Comment (lutter) [inline graphql/tests/query.rs]: Looks like this test needs just a little more work ;)

File: graph/src/data/query/result.rs
Comment (lutter) [inline graph/src/data/query/result.rs:191]: We should really nuke `extensions` since it is always `None`

**Review by lutter — CHANGES_REQUESTED** [review]

File: store/postgres/src/metadata.rs
Comment (lutter) [inline store/postgres/src/metadata.rs:795]: You also need to clamp the range of precious errors, otherwise, if you insert an error with range `[1,)`, you will never be able to insert another error for that subgraph.

And since `block_number = lower(block_range)`, do we even need that column?

**Review by lutter — CHANGES_REQUESTED** [review]

File: store/postgres/src/metadata.rs
Comment (lutter) [reply store/postgres/src/metadata.rs:795]: Since `on_conflict_do_nothing()` can mask problems silently, we need tests that check the various combinations of storing errors and that they work; at a minimum, it should check
* storing the same error twice stores only one error
* storing the same error at two different blocks stores two errors
* reverting correctly reverts errors

And it would be really nice if errors had a `created_at` column to make it easier to find new errors.

File: graph/src/components/store.rs
Comment (leoyvens) [reply graph/src/components/store.rs]: Good news is I implemented this without the persistent data structure, which really was bad for performance. Bad news is the code touched some tricky logic in the entity cache and is not super robust since it requires explicitly entering and exiting the handler. But I added and end-to-end test and plenty of asserts.

File: core/src/subgraph/instance_manager.rs
Comment (leoyvens) [reply core/src/subgraph/instance_manager.rs]: Changed.

File: runtime/wasm/src/module/mod.rs
Comment (leoyvens) [reply runtime/wasm/src/module/mod.rs]: Changed to the previous message.

File: core/src/subgraph/instance_manager.rs
Comment (leoyvens) [reply core/src/subgraph/instance_manager.rs:35]: Yhea graph node itself, updated the comment.

File: store/postgres/src/network_store.rs
Comment (leoyvens) [reply store/postgres/src/network_store.rs:55]: We agreed that this is ok.

File: store/postgres/src/network_store.rs
Comment (leoyvens) [reply store/postgres/src/network_store.rs]: Yes it wasn't a great idea to use anyhow, I switched this and a bunch of other places to StoreError.

File: store/postgres/src/metadata.rs
Comment (leoyvens) [reply store/postgres/src/metadata.rs]: Done.

File: graph/src/data/query/error.rs
Comment (leoyvens) [reply graph/src/data/query/error.rs]: Super weird, refactored this away.

File: graphql/src/schema/api.rs
Comment (leoyvens) [reply graphql/src/schema/api.rs]: That is nice! Moved it there.

File: graph/src/data/query/result.rs
Comment (leoyvens) [reply graph/src/data/query/result.rs]: Yay done

File: store/postgres/src/metadata.rs
Comment (leoyvens) [reply store/postgres/src/metadata.rs:793]: Dropped `block_number`, added `created_at`, added some tests that the conflict do nothing works.

**Comment by leoyvens** [comment]
@lutter I believe I've addressed all your comments, happy reviewing and rebasing!

File: graph/src/data/query/error.rs
Comment (lutter) [inline graph/src/data/query/error.rs]: Leftover from trying things out? Should probably be removed

File: store/postgres/migrations/2020-12-7-190800_subgraph_error_changes/up.sql
Comment (lutter) [inline store/postgres/migrations/2020-12-7-190800_subgraph_error_changes/up.sql:9]: The data type should be `timstamptz` so that [the timezone gets stored](https://www.postgresql.org/docs/9.6/datatype-datetime.html), too. Just `timestamp` defaults to `without time zone`. And I think you can just say `default now()` if we use a type with time zone.

File: store/postgres/src/metadata.rs
Comment (lutter) [inline store/postgres/src/metadata.rs:918]: I am not against this, but I thought we put tests that require external components like the store into `tests/`?

File: store/postgres/src/entities.rs
Comment (lutter) [inline store/postgres/src/entities.rs:776]: Is this necessary? There's an `on delete cascade` on the foreign key `subgraph_error.subgraph_id`

File: store/postgres/src/metadata.rs
Comment (lutter) [reply store/postgres/src/metadata.rs:918]: Actually, leave that here (fear of merge conflicts) - I can always move it later

**Review by lutter — APPROVED** [review]
Great! Once the tests pass, this is ready to merge.

File: store/postgres/src/entities.rs
Comment (leoyvens) [reply store/postgres/src/entities.rs]: Oh that's why we could get away without it. Removed it.

File: store/postgres/migrations/2020-12-7-190800_subgraph_error_changes/up.sql
Comment (leoyvens) [reply store/postgres/migrations/2020-12-7-190800_subgraph_error_changes/up.sql]: Done.

**Pull request merged**
